File: sdks/go/fireweave/runtime.go
package fireweave

import (
	"context"
	"errors"
	"sync"
)

// State is the runtime lifecycle state (docs/architecture.md).
type State string

const (
	StateUninitialized State = "UNINITIALIZED"
	StateInitializing  State = "INITIALIZING"
	StateReady         State = "READY"
	StateStale         State = "STALE"
	StateError         State = "ERROR"
	StateFatal         State = "FATAL"
	StateShutdown      State = "SHUTDOWN"
)

// Config configures a Runtime.
type Config struct {
	// Limits bounds evaluation contexts; zero fields fall back to
	// DefaultLimits.
	Limits Limits
	// RequireTargetingKey rejects evaluations whose merged context has no
	// targeting key (TARGETING_KEY_MISSING at the OpenFeature boundary).
	RequireTargetingKey bool
	// GlobalContext is an optional lowest-priority context layer merged
	// beneath every per-call context. It is copied at construction.
	GlobalContext EvaluationContext
}

// Runtime is the shared Fireweave engine: it owns the lifecycle state
// machine, gates evaluations on state, merges and validates contexts, and
// delegates resolution to a BackendAdapter.
//
// Concurrency guarantees:
//   - All methods are safe for concurrent use by multiple goroutines.
//   - Initialize is serialized; concurrent callers observe a single adapter
//     initialization (subsequent calls on a READY runtime are no-ops).
//   - Evaluate takes only a read lock on lifecycle state; adapter Resolve
//     calls run without any runtime lock, so evaluations proceed in
//     parallel.
//   - Shutdown is idempotent. Evaluations racing with Shutdown either
//     complete against the still-open adapter or observe SHUTDOWN and
//     return an AlreadyClosed default decision. After Shutdown returns, all
//     new evaluations return AlreadyClosed.
type Runtime struct {
	adapter             BackendAdapter
	limits              Limits
	requireTargetingKey bool
	globalCtx           EvaluationContext

	mu       sync.RWMutex // guards state + fatalErr
	initMu   sync.Mutex   // serializes Initialize
	state    State
	fatalErr *Error
}

// NewRuntime constructs a runtime in the UNINITIALIZED state.
func NewRuntime(adapter BackendAdapter, cfg Config) *Runtime {
	return &Runtime{
		adapter:             adapter,
		limits:              cfg.Limits.withDefaults(),
		requireTargetingKey: cfg.RequireTargetingKey,
		globalCtx:           cfg.GlobalContext.Copy(),
		state:               StateUninitialized,
	}
}

// State returns the current lifecycle state.
func (r *Runtime) State() State {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.state
}

// Adapter returns the backend adapter the runtime was built with.
func (r *Runtime) Adapter() BackendAdapter { return r.adapter }

// InitError returns the fatal initialization error, if any.
func (r *Runtime) InitError() *Error {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.fatalErr
}

// Initialize drives UNINITIALIZED → INITIALIZING → READY. Configuration
// failures land in FATAL; other failures land in ERROR (retryable via a
// subsequent Initialize). Initialize after Shutdown returns AlreadyClosed.
// Concurrent calls are serialized and the adapter is initialized once.
func (r *Runtime) Initialize(ctx context.Context) error {
	r.initMu.Lock()
	defer r.initMu.Unlock()

	r.mu.Lock()
	switch r.state {
	case StateShutdown:
		r.mu.Unlock()
		return NewError(KindAlreadyClosed, "", nil)
	case StateReady, StateStale:
		r.mu.Unlock()
		return nil
	case StateFatal:
		err := r.fatalErr
		r.mu.Unlock()
		return err
	}
	r.state = StateInitializing
	r.mu.Unlock()

	err := r.adapter.Initialize(ctx)

	r.mu.Lock()
	defer r.mu.Unlock()
	if err != nil {
		fwErr := asFireweaveError(err, KindConfiguration)
		if fwErr.Kind == KindConfiguration {
			r.state = StateFatal
			r.fatalErr = fwErr
		} else {
			r.state = StateError
		}
		return fwErr
	}
	r.state = StateReady
	return nil
}

// Evaluate resolves a flag. It never returns an error: failures surface as
// a Decision carrying the caller default, Reason ERROR, and a typed Error
// (OpenFeature default semantics).
func (r *Runtime) Evaluate(ctx context.Context, req ResolveRequest) Decision {
	switch r.State() {
	case StateReady, StateStale:
		// proceed
	case StateShutdown:
		return ErrorDecision(req.DefaultValue, NewError(KindAlreadyClosed, "", nil), nil)
	default:
		return ErrorDecision(req.DefaultValue, NewError(KindNotReady, "", nil), nil)
	}

	merged := MergeContexts(r.globalCtx, req.Context)
	if vErr := ValidateContext(merged, r.limits, r.requireTargetingKey); vErr != nil {
		return ErrorDecision(req.DefaultValue, vErr, nil)
	}

	req.Context = merged
	d := r.adapter.Resolve(ctx, req)
	if d.Error != nil {
		if d.Metadata == nil {
			d.Metadata = map[string]any{}
		}
		if _, ok := d.Metadata[MetaErrorKind]; !ok {
			d.Metadata[MetaErrorKind] = string(d.Error.Kind)
		}
		d.Reason = ReasonError
	}
	return d
}

// MarkStale transitions READY → STALE (adapter signals that flag data is
// being served from an aging cache). Evaluations continue in STALE.
func (r *Runtime) MarkStale() {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.state == StateReady {
		r.state = StateStale
	}
}

// Flush drains adapter-buffered telemetry when the adapter supports it.
func (r *Runtime) Flush(ctx context.Context) error {
	if sink, ok := r.adapter.(TelemetrySink); ok {
		return sink.FlushTelemetry(ctx)
	}
	return nil
}

// Shutdown transitions to SHUTDOWN and closes the adapter exactly once.
// Additional calls are no-ops returning nil (idempotent shutdown).
func (r *Runtime) Shutdown(ctx context.Context) error {
	r.mu.Lock()
	if r.state == StateShutdown {
		r.mu.Unlock()
		return nil
	}
	r.state = StateShutdown
	r.mu.Unlock()

	if err := r.adapter.Close(ctx); err != nil {
		return asFireweaveError(err, KindInternal)
	}
	return nil
}

// asFireweaveError coerces an arbitrary error into a typed *Error,
// defaulting to the given kind. The original error is wrapped as the cause;
// its text is never copied into the safe message.
func asFireweaveError(err error, fallback ErrorKind) *Error {
	var fwErr *Error
	if errors.As(err, &fwErr) {
		return fwErr
	}
	return NewError(fallback, "", err)
}

File: sdks/python/src/fireweave/runtime.py
"""Shared Fireweave runtime: lifecycle state machine + evaluation pipeline.

Lifecycle (docs/architecture.md):

    UNINITIALIZED -> INITIALIZING -> READY <-> STALE
                                   \\-> ERROR -> READY (recovery)
                                   \\-> FATAL
    any state ----------------------> SHUTDOWN (terminal, idempotent)

The runtime is the single synchronization point: a re-entrant lock guards
state transitions; evaluation reads state without blocking other evaluations.
Evaluation NEVER raises — every failure becomes a default-valued
:class:`~fireweave.decision.Decision` carrying the canonical error kind.
"""

from __future__ import annotations

import enum
import json
import threading
from typing import Any, Dict, Optional

from .adapters.base import BackendAdapter, FlagResolution
from .config import FireweaveConfig
from .context import EvaluationContext, merge_contexts, validate_context
from .decision import Decision, Reason
from .errors import (
    AlreadyClosedError,
    FireweaveError,
    FlagNotFoundError,
    InternalError,
    NotReadyError,
    TypeMismatchError,
    UnsupportedCapabilityError,
    FLAG_METADATA_ERROR_KIND_KEY,
)
from .types import FlagMetadata, FlagType

__all__ = ["LifecycleState", "FireweaveRuntime", "EvaluationOptions"]


class LifecycleState(enum.Enum):
    UNINITIALIZED = "UNINITIALIZED"
    INITIALIZING = "INITIALIZING"
    READY = "READY"
    STALE = "STALE"
    ERROR = "ERROR"
    FATAL = "FATAL"
    SHUTDOWN = "SHUTDOWN"

    @property
    def wire_name(self) -> str:
        """Provider-state name used by contracts fixtures."""
        return {
            LifecycleState.UNINITIALIZED: "NOT_READY",
            LifecycleState.INITIALIZING: "NOT_READY",
            LifecycleState.READY: "READY",
            LifecycleState.STALE: "STALE",
            LifecycleState.ERROR: "ERROR",
            LifecycleState.FATAL: "FATAL",
            LifecycleState.SHUTDOWN: "CLOSED",
        }[self]


class EvaluationOptions:
    """Per-invocation evaluation options.

    ``include_payload`` attaches the flag payload as ``fireweave.payload``
    metadata. ``send_exposure`` requests OF-path exposure emission; the
    phase-one Python PostHog adapter uses side-effect-free snapshot reads, so
    ``True`` is accepted for API portability but does not emit vendor
    ``$feature_flag_called`` on evaluate (use ``exposures.record`` /
    ``exposures.flush``). Default ``False`` matches the ratified phase-one
    side-effect-free evaluate contract (ADR-0001 errata / Agent M H-4).
    """

    __slots__ = ("include_payload", "send_exposure")

    def __init__(
        self, *, include_payload: bool = False, send_exposure: bool = False
    ) -> None:
        self.include_payload = include_payload
        self.send_exposure = send_exposure


_PY_TYPES = {
    FlagType.BOOLEAN: bool,
    FlagType.STRING: str,
}


def _check_type(flag_type: FlagType, value: Any) -> Any:
    """Strict type check with the ratified numeric coercion rules.

    - bool is never accepted for integer/float (Python bool subclasses int).
    - int -> float coercion is allowed; float -> int is a TypeMismatch even
      for integral floats (2.0 requested as integer fails).
    """
    if flag_type is FlagType.BOOLEAN:
        if isinstance(value, bool):
            return value
    elif flag_type is FlagType.STRING:
        if isinstance(value, str):
            return value
    elif flag_type is FlagType.INTEGER:
        if isinstance(value, int) and not isinstance(value, bool):
            return value
    elif flag_type is FlagType.FLOAT:
        if isinstance(value, (int, float)) and not isinstance(value, bool):
            return float(value)
    elif flag_type is FlagType.OBJECT:
        if isinstance(value, (dict, list)):
            return value
    raise TypeMismatchError()


class FireweaveRuntime:
    """Owns lifecycle, context layering, and the evaluation pipeline."""

    def __init__(
        self,
        adapter: BackendAdapter,
        config: Optional[FireweaveConfig] = None,
        *,
        global_context: Optional[EvaluationContext] = None,
    ) -> None:
        self._adapter = adapter
        self._config = config or FireweaveConfig()
        self._global_context = global_context
        self._client_context: Optional[EvaluationContext] = None
        self._state = LifecycleState.UNINITIALIZED
        self._lock = threading.RLock()
        self._init_error: Optional[FireweaveError] = None

    # -- lifecycle ---------------------------------------------------------

    @property
    def state(self) -> LifecycleState:
        with self._lock:
            return self._state

    @property
    def config(self) -> FireweaveConfig:
        return self._config

    @property
    def adapter(self) -> BackendAdapter:
        return self._adapter

    def initialize(self, *, backend_required: bool = False) -> None:
        """Transition UNINITIALIZED -> READY; FATAL on configuration failure.

        Raises the underlying :class:`FireweaveError` so callers (e.g. the
        OpenFeature provider ``initialize`` hook) can propagate it; the
        runtime state is updated first so later evaluations degrade safely.
        """
        with self._lock:
            if self._state is LifecycleState.SHUTDOWN:
                raise AlreadyClosedError()
            if self._state is LifecycleState.READY:
                return
            self._state = LifecycleState.INITIALIZING
        try:
            self._config.validate(backend_required=backend_required)
            self._adapter.initialize()
        except FireweaveError as exc:
            with self._lock:
                self._state = LifecycleState.FATAL
                self._init_error = exc
            raise
        except Exception as exc:  # vendor exception: wrap, preserve cause
            wrapped = InternalError("initialization failed")
            wrapped.__cause__ = exc
            with self._lock:
                self._state = LifecycleState.FATAL
                self._init_error = wrapped
            raise wrapped from exc
        with self._lock:
            self._state = LifecycleState.READY
            self._init_error = None

    def lifecycle_gate(self) -> Optional[FireweaveError]:
        """Extension-call lifecycle gate (ruling 17, Go/Java model).

        READY/STALE pass (``None``); after shutdown the gate is
        ``AlreadyClosed``; any pre-ready state degrades with
        ``UnsupportedCapability``. Callers convert the returned error into a
        structured result — extension APIs never raise for lifecycle reasons.
        """
        state = self.state
        if state in (LifecycleState.READY, LifecycleState.STALE):
            return None
        if state is LifecycleState.SHUTDOWN:
            return AlreadyClosedError()
        return UnsupportedCapabilityError()

    def mark_stale(self) -> None:
        with self._lock:
            if self._state is LifecycleState.READY:
                self._state = LifecycleState.STALE

    def force_state(self, state: LifecycleState) -> None:
        """Test/conformance hook: pin the lifecycle state."""
        with self._lock:
            self._state = state

    def shutdown(self, timeout_ms: Optional[int] = None) -> None:
        """Deterministic, idempotent shutdown; never raises."""
        with self._lock:
            if self._state is LifecycleState.SHUTDOWN:
                return
            self._state = LifecycleState.SHUTDOWN
        try:
            self._adapter.shutdown(timeout_ms or self._config.shutdown_timeout_ms)
        except Exception:
            # Shutdown must not throw; adapter errors are swallowed by design.
            pass

    # -- context layering ---------------------------------------------------

    def set_global_context(self, context: Optional[EvaluationContext]) -> None:
        with self._lock:
            self._global_context = context

    def set_client_context(self, context: Optional[EvaluationContext]) -> None:
        with self._lock:
            self._client_context = context

    def merged_context(
        self, invocation: Optional[EvaluationContext]
    ) -> EvaluationContext:
        with self._lock:
            global_ctx, client_ctx = self._global_context, self._client_context
        return merge_contexts(global_ctx, client_ctx, invocation)

    # -- evaluation pipeline -------------------------------------------------

    def evaluate(
        self,
        flag_key: str,
        flag_type: FlagType,
        default_value: Any,
        invocation_context: Optional[EvaluationContext] = None,
        options: Optional[EvaluationOptions] = None,
    ) -> Decision:
        """Evaluate a flag. Never raises; failures return the default."""
        options = options or EvaluationOptions()

        state = self.state
        if state is LifecycleState.SHUTDOWN:
            return self._error_decision(default_value, AlreadyClosedError())
        if state in (LifecycleState.UNINITIALIZED, LifecycleState.INITIALIZING):
            return self._error_decision(default_value, NotReadyError())
        if state is LifecycleState.FATAL:
            err = self._init_error or NotReadyError()
            return self._error_decision(default_value, err)

        merged = self.merged_context(invocation_context)
        try:
            validate_context(
                merged,
                limits=self._config.limits,
                reserved_keys=self._config.reserved_attribute_keys,
                require_targeting_key=self._config.require_targeting_key,
            )
        except FireweaveError as exc:
            return self._error_decision(default_value, exc, merged)

        try:
            resolution = self._adapter.resolve(flag_key, merged)
        except FireweaveError as exc:
            return self._error_decision(default_value, exc, merged)
        except Exception as exc:
            wrapped = InternalError("evaluation failed")
            wrapped.__cause__ = exc
            return self._error_decision(default_value, wrapped, merged)

        return self._decision_from_resolution(
            resolution, flag_type, default_value, merged, options, state
        )

    # -- helpers -------------------------------------------------------------

    def _decision_from_resolution(
        self,
        resolution: FlagResolution,
        flag_type: FlagType,
        default_value: Any,
        merged: EvaluationContext,
        options: EvaluationOptions,
        state: LifecycleState,
    ) -> Decision:
        if not resolution.matched:
            return Decision(
                value=default_value,
                variant=None,
                reason=Reason.DEFAULT,
                resolved_context=self._resolved_context_dict(merged),
            )

        try:
            value = _check_type(flag_type, resolution.value)
        except TypeMismatchError as exc:
            return self._error_decision(default_value, exc, merged)

        if resolution.fireweave_reason:
            reason = resolution.fireweave_reason
        elif not resolution.enabled:
            reason = Reason.DISABLED
        elif resolution.from_cache or state is LifecycleState.STALE:
            reason = Reason.STALE
        else:
            reason = Reason.TARGETING_MATCH

        metadata: FlagMetadata = {}
        if resolution.version is not None:
            metadata["fireweave.flagVersion"] = resolution.version
        # Detailed enrichment: only when the vendor supplied both a flag id
        # and a matched-condition index (see conformance notes).
        if (
            resolution.vendor_flag_id is not None
            and resolution.condition_index is not None
            and resolution.reason_code is not None
        ):
            metadata["fireweave.vendorFlagId"] = resolution.vendor_flag_id
            metadata["fireweave.reasonCode"] = resolution.reason_code
        if options.include_payload and resolution.payload is not None:
            metadata["fireweave.payload"] = json.dumps(
                resolution.payload, sort_keys=True, separators=(",", ":")
            )
        if resolution.from_cache:
            metadata["fireweave.fromCache"] = True
        metadata.update(resolution.extra_metadata)

        return Decision(
            value=value,
            variant=resolution.variant,
            reason=reason,
            flag_metadata=metadata,
            resolved_context=self._resolved_context_dict(merged),
        )

    @staticmethod
    def _resolved_context_dict(merged: EvaluationContext) -> Dict[str, Any]:
        out: Dict[str, Any] = {}
        if merged.targeting_key is not None:
            out["targetingKey"] = merged.targeting_key
        plain = merged.plain_attributes
        if plain:
            out["attributes"] = plain
        return out

    def _error_decision(
        self,
        default_value: Any,
        error: FireweaveError,
        merged: Optional[EvaluationContext] = None,
    ) -> Decision:
        metadata: FlagMetadata = {FLAG_METADATA_ERROR_KIND_KEY: error.kind.value}
        if isinstance(error, FlagNotFoundError) and error.quota_limited:
            metadata["fireweave.quotaLimited"] = True
        return Decision(
            value=default_value,
            variant=None,
            reason=Reason.ERROR,
            error_code=error.openfeature_error_code,
            error_message=error.message,
            error_kind=error.kind,
            flag_metadata=metadata,
            resolved_context=(
                self._resolved_context_dict(merged) if merged is not None else None
            ),
        )
